text/0000-deprecate-globals-resolver.md

```diff
@@ -0,0 +1,122 @@
+- Start Date: 2018-05-08
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Deprecate all use of:
+
+- Ember Globals Resolver (looks up a class via a global namespace such as "App")
+- Creation of a Global Namespace (`var App = Ember.Namespace.create();`)
+- Ember.TEMPLATES array
+- &lt;script type="text/handlebars" data-template-name="path/to/template"&gt;
+
+Use of any of the above should trigger a deprecation warning, with a target
+of version 4.0
+
+# Motivation
+
+Over the past years we have transitioned to using Ember-CLI as the main way
+to compile Ember apps. The globals resolver is a holdover and primarily
+facilitates use of Ember without Ember-CLI.
+
+# The Globals Resolver
+
+For those who are not aware, the globals resolver is available via `@ember/globals-resolver` or
+`Ember.DefaultResolver`. For more information, see the
+[api](https://www.emberjs.com/api/ember/release/classes/GlobalsResolver/properties).
+Using it looks like the following:
+
+```js
+// app.js
+var App = Ember.Application.create();
+
+App.Router.map(function() {
+  this.route('about');
+});
+
+App.AboutRoute = Ember.Route.extend({
+  model: function() {
+    return ['red', 'yellow', 'blue'];
+  }
+});
+```
+
+```html
+// index.html
+<script type="text/x-handlebars" data-template-name="about">
+  <ul>
+    {{#each model as |item|}}
+      <li>{{item}}</li>
+    {{/each}}
+  </ul>
+</script>
+```
+
+# Transition Path
+
+Primarily, the transition path is to recommend using Ember-CLI.
+
+One small detail required to implement this RFC: ember-cli's own default resolver,
+[ember-resolver](https://github.com/ember-cli/ember-resolver)
+currently still extends from the globals resolver.
+In order to implement this RFC, the ember-cli resolver will need to be changed
+so that it does *not* extend from the globals resolver, or otherwise ember-cli users
+will get a deprecation warning as well.
```

Comment:
We can't remove the global resolver base class for the ember-cli classic resolver, that would be a breaking change.

I can think of two options:

**Simple option**

Have the global resolver log a deprecation message at runtime. When `resolve` or any other method on the resolver is hit, deprecation message. Then we remove a) the global resolver and b) ember-cli classic resolver's extension in 4.0.

**Complex, faster option**

In the ember-cli classic resolver, deprecate any runtime calls where there is fallback to the globals mode resolver. This would be a deprecation in ember-cli's resolver. We could bump a major version of ember-cli-resolver removing the base class and release it in ember-cli after an LTS of ember-cli.

Then separately (or concurrently with some cleverness) we could deprecate any use of the ember-cli globals resolver.

Reply:
> In the ember-cli classic resolver, deprecate any runtime calls where there is fallback to the globals mode resolver. This would be a deprecation in ember-cli's resolver. We could bump a major version of ember-cli-resolver removing the base class and release it in ember-cli after an LTS of ember-cli.

This was what I was planning on doing.

Reply:
Unless there are any objections, I'll put this in the RFC?

Reply:
Yes, sounds good to update with that.